unittests/runtime/Metadata.cpp

```diff
@@ -21,127 +21,73 @@
-  if (barrier->numThreadsWaiting < barrier->count) {
-    // Put the thread to sleep.
-    if (pthread_cond_wait(&barrier->cond, &barrier->mutex) != 0) {
-      return -1;
-    }
-    if (pthread_mutex_unlock(&barrier->mutex) != 0) {
-      return -1;
-    }
-    return 0;
-  } else {
-    // Reset thread count.
-    barrier->numThreadsWaiting = 0;
-
-    // Wake up all threads.
-    if (pthread_cond_broadcast(&barrier->cond) != 0) {
-      return -1;
-    }
-    if (pthread_mutex_unlock(&barrier->mutex) != 0) {
-      return -1;
-    }
-    return PTHREAD_BARRIER_SERIAL_THREAD;
-  }
-}
-#endif
+#include <thread>
+#include <condition_variable>
 
 using namespace swift;
 
 // Race testing.
 
 template <typename T>
-struct RaceArgs {
+struct RaceThreadContext {
   std::function<T()> code;
-  pthread_barrier_t *go;
+  T result;
+
+  unsigned numThreads;
+  unsigned &numThreadsReady;
+  std::mutex &sharedMutex;
+  std::condition_variable &start_condition;
 };
 
-void *RaceThunk(void *vargs) {
-  RaceArgs<void*> *args = static_cast<RaceArgs<void*> *>(vargs);
-  // Signal ready. Wait for go.
-  pthread_barrier_wait(args->go);
-  return args->code();
+template <typename T>
+void RaceThunk(RaceThreadContext<T> ctx) {
+  // update shared state
+  std::unique_lock<std::mutex> lk(ctx.sharedMutex);
+  ++ctx.numThreadsReady;
+  bool isLastThread = ctx.numThreadsReady == ctx.numThreads;
+
+  // wait until the rest of the thunks are ready
+  ctx.start_condition.wait(lk, [&ctx]{ // waiting releases the lock
+    return ctx.numThreadsReady == ctx.numThreads;
+  });
+  lk.unlock();
+
+  // The last thread will signal the condition_variable to kick off the rest
+  // of the waiting threads to start.
+  if (isLastThread) ctx.start_condition.notify_all();
+
+  ctx.result = ctx.code();
 }
 
 /// RaceTest(code) runs code in many threads simultaneously, 
 /// and returns a vector of all returned results.
-template <typename T, int NumThreads = 64>
-std::vector<T> 
-RaceTest(std::function<T()> code)
-{
-  const unsigned threadCount = NumThreads;
-
-  pthread_barrier_t go;
-  pthread_barrier_init(&go, nullptr, threadCount);
-
-  // Create the threads.
-  pthread_t threads[threadCount];
-  std::vector<RaceArgs<T>> args(threadCount, {code, &go});
-
-  for (unsigned i = 0; i < threadCount; i++) {
-    pthread_create(&threads[i], nullptr, &RaceThunk, &args[i]);
+template <typename T, unsigned NumThreads = 64>
+std::vector<T> RaceTest(std::function<T()> code) {
+  unsigned numThreadsReady = 0;
+  std::mutex sharedMutex;
+  std::condition_variable start_condition;
+
+  // Create the vector + contexts.
+  std::vector<std::thread> threads(NumThreads);
+  std::vector<RaceThreadContext<T>> contexts(NumThreads, {
+		  /*code*/ code,
+		  /*result*/ nullptr,
+		  NumThreads,
+		  numThreadsReady,
+		  sharedMutex,
+		  start_condition});
+
+  // Create all the threads
+  for (unsigned i = 0; i < NumThreads; i++) {
+    threads.emplace_back(std::thread(RaceThunk<T>, contexts[i]));
```

Comment:
This combination of sized constructor and `emplace_back()` looks wrong. I think it creates a vector of `2*NumThreads` elements, with the first half being default-initialized and the second half being the desired contents.

Reply:
I'm sorry! I'm still learning cpp, my bad 😬
How should I update the vector initializer to not fill in default values? I looked at http://en.cppreference.com/w/cpp/container/vector/vector but it doesn't seem easy to have the `Allocator` not insert values. Should I switch from `emplace_back`ing to assigning to `threads[i]`?

Reply:
For this code simply assigning to `threads[i]` should be fine. Or create an empty vector and `push_back` or `emplace_back` onto it, like the old pthreads code did. Both of these might have performance downsides, but we don't care in this code.

This stackoverflow answer has a good overview of the combinations that work and their performance effects. 
https://stackoverflow.com/a/32200517/881456